src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -298,29 +298,23 @@ abstract class BTypes {
    * referring to BTypes.
    */
   sealed trait BType {
-    final override def toString: String = this match {
-      case UNIT   => "V"
-      case BOOL   => "Z"
-      case CHAR   => "C"
-      case BYTE   => "B"
-      case SHORT  => "S"
-      case INT    => "I"
-      case FLOAT  => "F"
-      case LONG   => "J"
-      case DOUBLE => "D"
-      case ClassBType(internalName) => "L" + internalName + ";"
-      case ArrayBType(component)    => "[" + component
-      case MethodBType(args, res)   => "(" + args.mkString + ")" + res
-    }
+
+    final override def toString: String = descriptor
 
     /**
-     * @return The Java descriptor of this type. Examples:
-     *  - int: I
-     *  - java.lang.String: Ljava/lang/String;
-     *  - int[]: [I
-     *  - Object m(String s, double d): (Ljava/lang/String;D)Ljava/lang/Object;
-     */
-    final def descriptor = toString
+      * @return The Java descriptor of this type. Examples:
+      *  - int: I
+      *  - java.lang.String: Ljava/lang/String;
+      *  - int[]: [I
+      *  - Object m(String s, double d): (Ljava/lang/String;D)Ljava/lang/Object;
+      */
+    def descriptor : String
+
+    /**
+      * generally used by descriptor to help build the descriptor. Allows several descriptors to be build from the same stringbuilder
+      * @param sb
+      */
+    private[BTypes] def buildDescriptor(sb: java.lang.StringBuilder): Unit
```

Comment:
Not sure what you are getting at
As I see it you either use a pattern match or virtual dispatch
Getting run of the 'this match { case ...} ' pattern in the code aids encapsulation, safety and performance IMHO

Why an overload? They are behaviourly different?

I am turn as to whether the buildDescriptor should return the descriptor  or generate i as a side effect. Either way there are 2 effects, and I could not come up with a a better name before my attention moved